dynatrace/stackstate_checks/dynatrace/dynatrace.py

```diff
@@ -0,0 +1,506 @@
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=instance_info.instance_tags, message=msg)
+        except EventLimitReachedException as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.WARNING, tags=instance_info.instance_tags,
+                               message=str(e))
+        except Exception as e:
+            self.log.exception(str(e))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=instance_info.instance_tags,
+                               message=str(e))
+
+    def _process_topology(self, instance_info):
+        """
+        Collects components and relations for each component type from dynatrace smartscape topology API
+        """
+        start_time = datetime.now()
+        self.log.debug("Starting the collection of topology")
+        for component_type, path in TOPOLOGY_API_ENDPOINTS.items():
+            endpoint = self._get_endpoint(instance_info.url, path)
+            params = {"relativeTime": instance_info.relative_time}
+            response = self._get_dynatrace_json_response(instance_info, endpoint, params)
+            self._collect_topology(response, component_type, instance_info)
+        end_time = datetime.now()
+        time_taken = end_time - start_time
+        self.log.info("Collected %d topology entities.", len(dynatrace_entities_cache))
+        self.log.debug("Time taken to collect the topology is: %d seconds" % time_taken.total_seconds())
+
+    def _collect_topology(self, response, component_type, instance_info):
+        """
+        Process each component type and map those with specific data
+        :param response: Response of each component type endpoint
+        :param component_type: Component type
+        :param instance_info: instance configuration
+        :return: create the component on stackstate API
+        """
+        for item in response:
+            item = self._clean_unsupported_metadata(item)
+            dynatrace_component = DynatraceComponent(item, strict=False)
+            dynatrace_component.validate()
+            data = {}
+            external_id = dynatrace_component.entityId
+            identifiers = [Identifiers.create_custom_identifier("dynatrace", external_id)]
+            if component_type == "host":
+                host_identifiers = self._get_host_identifiers(dynatrace_component)
+                identifiers.extend(host_identifiers)
+            # derive useful labels from dynatrace tags
+            tags = self._get_labels(dynatrace_component)
+            tags.extend(instance_info.instance_tags)
+            data.update(item)
+            self._filter_item_topology_data(data)
+            data.update({
+                "identifiers": identifiers,
+                "tags": tags,
+                "domain": instance_info.domain,
+                "environment": instance_info.environment,
```

Comment:
This is called `environments` now and it's value should be a list.

Reply:
done